python/benchmark/gen_data_distributed.py

```diff
@@ -144,6 +142,97 @@ def make_blobs_udf(iter: Iterable[pd.DataFrame]) -> Iterable[pd.DataFrame]:
+            make_low_rank_matrix, ["n_samples", "n_features"]
+        )
+        # must replace the None to the correct type
+        params["random_state"] = int
+        return params
+
+    def gen_dataframe(self, spark: SparkSession) -> Tuple[DataFrame, List[str]]:
+        dtype = self.dtype
+        params = self.extra_params
+
+        if "random_state" not in params:
+            # for reproducible dataset.
+            params["random_state"] = 1
+
+        print(f"Passing {params} to make_low_rank_matrix")
+
+        rows = self.num_rows
+        cols = self.num_cols
+        assert self.args is not None
+        num_partitions = self.args.output_num_files
+
+        # Set num_partitions to Spark's default if output_num_files is not provided.
+        if num_partitions is None:
+            num_partitions = spark.sparkContext.defaultParallelism
+
+        generator = np.random.RandomState(params["random_state"])
+        n = min(rows, cols)
+        # If params not provided, set to defaults.
+        effective_rank = params.get("effective_rank", 10)
+        tail_strength = params.get("tail_strength", 0.5)
+
+        partition_sizes = [rows // num_partitions] * num_partitions
+        partition_sizes[-1] += rows % num_partitions
+        # Check sizes to ensure QR decomp produces a matrix of the correct dimension.
+        assert partition_sizes[0] >= cols, (
+            f"Num samples per partition ({partition_sizes[0]}) must be >= num_features ({cols});"
+            f" decrease num_partitions from {num_partitions} to <= {rows // cols}"
+        )
+
+        # Generate U, S, V, the SVD decomposition of the output matrix.
+        # S and V are generated upfront, U is generated across partitions.
+        singular_ind = np.arange(n, dtype=dtype)
+        low_rank = (1 - params["tail_strength"]) * np.exp(
+            -1.0 * (singular_ind / params["effective_rank"]) ** 2
+        )
+        tail = tail_strength * np.exp(-0.1 * singular_ind / effective_rank)
+        s = np.identity(n) * (low_rank + tail)
+        # compute V upfront
+        v, _ = linalg.qr(
+            generator.standard_normal(size=(cols, n)),
+            mode="economic",
+            check_finite=False,
+        )
+
+        # UDF for distributed generation of U and the resultant product U*S*V.T
+        def make_matrix_udf(iter: Iterable[pd.Series]) -> Iterable[pd.DataFrame]:
+            for pdf in iter:
+                partition_index = pdf.iloc[0][0]
+                n_partition_rows = partition_sizes[partition_index]
+                u, _ = linalg.qr(
+                    generator.standard_normal(size=(n_partition_rows, n)),
+                    mode="economic",
+                    check_finite=False,
+                )
+                # Include partition-wise normalization to ensure overall unit norm.
+                u *= np.sqrt(1 / num_partitions)
+                mat = np.dot(np.dot(u, s), v.T).astype(dtype)
```

Comment:
Not sure if computing the intermediate `math` and then deleting `u` is more memory efficient than yielding `DataFrame(data=np.dot(u, sv_normed))`. I figure since they're the same shape the approaches will be similar